### ISSUE TITLE
[question] Conan Package does not define the library

### ISSUE THREAD

**Issue by melMass** [conan-io/conan#9358]
Hi

## Reading the TLDR first might be enough to grasp the issue, but here is a bit of context:

I've been using Conan a lot in the past year, and I'm converting all my project & most common external libs to it (conan-center is a very good companion btw).

As there are many ways to approach a package, and things evolved over time, I think I need a bit of a refresher.
I decided to switch all of my package to use the `cmake_find_package` & `virtualenv` combo to limit the number of files I have to maintain (and staying closer to pure CMake).

I'm not sure if it's related to switching to `cmake_find_package`, but I have a linking error to my package from my test_package (or any other package, I just tested to make sure.).

What is strange is that if I remove the `test_package`, the main package gets exported and inspecting its cache, my includes and library are there, in their respective folders: `include` & `lib`. I then noticed a warning from conan when building the test package: `Library mtb-core not found in package, might be system one`

Looking at the template test_package, I realized you were using `imports()` to copy the libraries for conan cache to the cmake build, isn't this handled by cmake_find? To be certain, I added `imports` into my test_package, and it's still not working but `mtb-core.dylib` is in the build folder under lib, I feel like I'm missing something obvious...

# TLDR
- Is `imports()` always required for libraries and headers?
- My package library isn't exported properly as I have the following warning from building my test_package:
  ```log
  -- Conan: Using autogenerated Findmtb-core.cmake
  -- Found mtb-core: 0.1 (found version "0.1")
  -- Library mtb-core not found in package, might be system one. <----- This is strange
  ```
   resulting in a link error: `ld: library not found for -lmtb-core`
    What could cause this? How does conan decide what is a lib from the packaged folder? (because it's there in my case)

### Relevant files

<details>
  <summary>
    the main <code>conanfile.py</code>
  </summary>

```python
from conans import ConanFile, CMake, tools
import os

class MtbCoreConan(ConanFile):
    name = "mtb-core"
    version = "0.1"
    license = "ArtisticLicensev2"
    author = "Mel Massadian"
    url = "-"
    description = "Core C++ Library"
    topics = "melMass"
    settings = "os", "compiler", "build_type", "arch"
    options = {"shared": [True, False]}
    default_options = {"shared": False}
    generators = ["cmake_find_package", "virtualenv"]
    exports_sources = ["src/*", "CMakeLists.txt"]
    requires = ["nlohmann_json/3.9.1", "zlib/1.2.11"]

    def build(self):
        cmake = CMake(self)
        cmake.configure()
        cmake.build()

    def package(self):
        self.copy("*.h", dst="include", src="src")
        self.copy("*.dylib*", dst="lib", keep_path=False)
        self.copy("*.a", dst="lib", keep_path=False)

   def package_info(self):
        # self.cpp_info.libs = tools.collect_libs(self)
        self.cpp_info.libs = ["mtb-core"]

```
</details>

<details>
  <summary>
    my test package <code>conanfile.py</code>
  </summary>

```python
import os
from conans import ConanFile, CMake, tools

class MtbCoreTestConan(ConanFile):
    settings = "os", "compiler", "build_type", "arch"
    generators = "cmake_find_package"

    def build(self):
        cmake = CMake(self)
        cmake.configure()
        cmake.build()

    def imports(self):
        self.copy("*.dll", dst="lib", src="bin")
        self.copy("*.dylib*", dst="lib", src="lib")
        self.copy("*.a", dst="lib", src="lib")
        self.copy("*.so*", dst="lub", src="lib")

    def test(self):
        if not tools.cross_building(self):
            os.chdir("bin")
            self.run(".%smtb-example" % os.sep)

```
</details>

- [x] I've read the [CONTRIBUTING guide](https://github.com/conan-io/conan/blob/develop/.github/CONTRIBUTING.md).

**Comment by melMass** [comment]
Wait.... I have no idea what is going on, but removing the library from conan's cache just solved the issue... and I don't need the imports either... So strange.
I'm leaving this open for a few days if it rings someone's bell or if I understand it myself in the meantime, I did regularly cleared the cache while doing my tests. I'll keep investigating.

**Comment by melMass** [comment]
I found the issue!
I removed every package from my cache, readded a few commented out parts of my main `CMakeFiles.txt` to find the issue:  
  
On macOS by default libraries are prepended with `lib`.  
In the CMake of `mtb-core` I'm defining the target like this to remove this prefix:

```cmake
set_target_properties(${PROJECT_NAME}
  PROPERTIES
  OUTPUT_NAME ${PROJECT_NAME}
  PREFIX ""
 )
```

This is what was breaking everything! Is there any way to set the target name somehow?

**Comment by lasote** [comment]
Hi! I strongly recommend using the new `CMakeDeps` generator in addition to the `CMakeToolchain` instead of the `cmake_find_package`. They are tagged as "experimental" but they are here to stay and replace all the cmake generators in Conan 2.0. You can find the documentation here: https://docs.conan.io/en/latest/reference/conanfile/tools/cmake.html

Said that I don't really understand your issue, the "lib" prefix shouldn't be an issue. If you could provide a package to test maybe we could help.

**Comment by melMass** [comment]
Thanks. 
I will try that, I'm a bit confused by CMakeToolchain, but I will experiment before asking further information

**Comment by melMass** [comment]
> Hi! I strongly recommend using the new `CMakeDeps` generator in addition to the `CMakeToolchain` instead of the `cmake_find_package`. 

Do you know recipes who already use these?

**Comment by lasote** [comment]
not yet in conan center but you can find multiple test using it:

https://github.com/conan-io/conan/tree/develop/conans/test/functional/toolchains/cmake

**Comment by melMass** [comment]
Thanks.
After a few tests I managed to understand a few things like:
- With these generators, the cmake configuration is made in the generate method, using either a `CMakeDeps` or `CMakeToolchains` class.
- Packaging seems very different, I still need to isolate a simple test case to understand this I think

What I don't understand is:
- The distinction between the `tools.CMake`, `CMakeDeps` and `CMakeToolchain`, more exactly when do you instantiate one over the other, it also does not seems to be connected to the generator used for instance [here](https://github.com/conan-io/conan/blob/develop/conans/test/functional/toolchains/cmake/test_cmake.py), the generator is set to `CMakeDeps` but the generate method uses `CMakeToolchain`, that being said from my local tests all of these combo works. 🤯

**Comment by lasote** [comment]
Hi! The packaging is exactly the same, I don't know what you mean.
The difference between all these tools you mention is:

- `CMakeToolchain`: Is a generator that writes a `conan_toolchain.cmake` file containing information about the current configuration, like build type, architecture, and some other things.  

- `CMakeDeps`: Is a generator that writes the `XXXConfig.cmake` files for each dependency, so you can use `find_package()` from a consumer project in the `CMakeLists.txt` file.

- `tools.CMake`: Is a very dummy build helper that calls cmake command line with the generated toolchain: `cmake /path/to/cmakelist_folder -DCMAKE_TOOLCHAIN_FILE=/path/to/generators_files/conan_toolchain` + `cmake --build .`

Why we did that component split? So if you have a local project (and your conanfile.py) you could do "conan install", it will generate both the `conan_toolchain.cmake` and the `XXXConfig.cmake` files and you could manually call cmake and it will do exactly the same than the `build()` method. Previously all the logic of the toolchain was "hardcoded" in the `CMake()` build helper (in python) so you cannot reproduce what conan was doing easily.

**Comment by melMass** [comment]
> Hi! The packaging is exactly the same, I don't know what you mean.

Ok ignore me ahah,  I think I've just barely used the possibilities of `cpp_info` and this doc revealed them to me

https://docs.conan.io/en/latest/reference/conanfile/tools/cmake/cmakedeps.html#properties
I'm not sure of when I should rely on conan's find_deps()

> Why we did that component split? 

Very interesting, ok I see the huge benefit of this approach.

---
I think I need to re-read the new documentation fully as I have missed a few important updates it seems.
I currently cannot isolate my original issue as I'm in the middle of a huge refactoring so I will close this issue and open a proper one with an isolated test case if I have a similar issue.

Thanks a lot for your answers @lasote
